File: dart/lib/src/bindings/create_record.dart
import '../constants/addresses.dart';
import '../constants/records.dart';
import '../domain/get_domain_address.dart';
import '../errors/sns_errors.dart';
import '../instructions/allocate_and_post_record_instruction.dart';
import '../instructions/instruction_types.dart';
import '../utils/serializers/serialize_record_content.dart';

/// Parameters for creating a record
class CreateRecordParams {
  const CreateRecordParams({
    required this.domain,
    required this.record,
    required this.content,
    required this.owner,
    required this.payer,
  });

  /// The domain under which the record will be created
  final String domain;

  /// The type of record to be created
  final Record record;

  /// The record content
  final String content;

  /// The address of the domain's owner
  final String owner;

  /// The address funding the record creation
  final String payer;
}

/// Creates a record for the specified domain. The record data will be
/// serialized in compliance with the SNS-IP 1 guidelines.
///
/// This mirrors js-kit/src/bindings/createRecord.ts
Future<TransactionInstruction> createRecord(CreateRecordParams params) async {
  final domainResult = await getDomainAddress(
    GetDomainAddressParams(
      domain: '${params.record.value}.'
          '${params.domain}',
      record: RecordVersion.v2,
    ),
  );

  var parentAddress = domainResult.parentAddress;
  final isSub = domainResult.isSub;

  if (isSub) {
    final parentResult = await getDomainAddress(
      GetDomainAddressParams(domain: params.domain),
    );
    parentAddress = parentResult.domainAddress;
  }

  if (parentAddress == null) {
    throw InvalidParentError('Parent could not be found');
  }

  final recordContent = serializeRecordContent(
    content: params.content,
    record: params.record,
  );

  final instruction = AllocateAndPostRecordInstruction(
    record: '\x02${params.record.value}',
    content: recordContent,
  )..setParams(
      programAddress: recordsProgramAddress,
      systemProgram: systemProgramAddress,
      splNameServiceProgram: nameProgramAddress,
      payer: params.payer,
      recordAddress: domainResult.domainAddress,
      domainAddress: parentAddress,
      domainOwner: params.owner,
      centralState: centralStateDomainRecords,
    );

  return instruction.build();
}

File: js/src/bindings/createRecord.ts
import { allocateAndPostRecord, SNS_RECORDS_ID } from "@bonfida/sns-records";
import { PublicKey } from "@solana/web3.js";

import { NAME_PROGRAM_ID } from "../constants";
import { InvalidParentError } from "../error";
import { serializeRecordContent } from "../record/serializeRecordContent";
import { Record, RecordVersion } from "../types/record";
import { getSnsDomainKeySync } from "../utils/getSnsDomainKeySync";
import { _parseSnsDomain } from "../utils/parseSnsDomain";

/**
 * Builds an instruction to create a record for a `.sns` domain or subdomain.
 *
 * @param domain Full `.sns` domain or subdomain name
 * @param record Record type
 * @param content Record content
 * @param owner Current owner of the domain
 * @param payer Fee payer for the instruction
 * @returns Transaction instruction.
 */
export const createRecord = (
  domain: string,
  record: Record,
  content: string,
  owner: PublicKey,
  payer: PublicKey,
) => {
  const trimmedDomain = _parseSnsDomain(domain);

  let { pubkey, parent, isSub } = getSnsDomainKeySync(
    `${record}.${trimmedDomain}`,
    RecordVersion.V2,
  );

  if (isSub) {
    parent = getSnsDomainKeySync(trimmedDomain).pubkey;
  }

  if (!parent) {
    throw new InvalidParentError("Parent could not be found");
  }

  const ix = allocateAndPostRecord(
    payer,
    pubkey,
    parent,
    owner,
    NAME_PROGRAM_ID,
    `\x02`.concat(record as string),
    serializeRecordContent(content, record),
    SNS_RECORDS_ID,
  );
  return ix;
};
